### ISSUE TITLE
Integration with pyimgui or nanogui -> Confirmed for pyimgui

### ISSUE THREAD

**Issue by mlviz** [glumpy/glumpy#174]
Would a pure python integration between glumpy and [pyimgui](https://github.com/swistakm/pyimgui) or [nanogui](https://github.com/wjakob/nanogui), using pyopengl and glfw, be possible? 

If not, what features should be added to glumpy to make it happen?

**Comment by mlviz** [comment]
After experimenting, I can confirm that pyimgui and glumpy can be combined from python:
![image](https://user-images.githubusercontent.com/45414538/49286533-6f8ba900-f468-11e8-8c80-0bc5c2415abe.png)

And here is the code:

```
# -*- coding: utf-8 -*-
import glfw
import OpenGL.GL as gl

import imgui
from imgui.integrations.glfw import GlfwRenderer

from glumpy.app import configuration
from glumpy import app, gloo, gl

vertex = """
    uniform float theta;
    attribute vec4 color;
    attribute vec2 position;
    varying vec4 v_color;
    void main()
    {
        float ct = cos(theta);
        float st = sin(theta);
        float x = 0.75* (position.x*ct - position.y*st);
        float y = 0.75* (position.x*st + position.y*ct);
        gl_Position = vec4(x, y, 0.0, 1.0);
        v_color = color;
    } """

fragment = """
  varying vec4 v_color;
  void main()
  {
      gl_FragColor = v_color;
  } """

def quad_init():
    # Build the program and corresponding buffers (with 4 vertices)
    quad = gloo.Program(vertex, fragment, count=4)
    # Upload data into GPU
    quad['color'] = [ (1,0,0,1), (0,1,0,1), (0,0,1,1), (1,1,0,1) ]
    quad['position'] = [ (-1,-1),   (-1,+1),   (+1,-1),   (+1,+1)   ]
    quad['theta'] = 0
    return quad

def main():
    window = impl_glfw_init()
    quad = quad_init()
    impl = GlfwRenderer(window)
    theta = 0.0
    
    while not glfw.window_should_close(window):
        glfw.poll_events()
        impl.process_inputs()

        imgui.new_frame()

        if imgui.begin_main_menu_bar():
            if imgui.begin_menu("File", True):

                clicked_quit, selected_quit = imgui.menu_item(
                    "Quit", 'Cmd+Q', False, True
                )

                if clicked_quit:
                    exit(1)

                imgui.end_menu()
            imgui.end_main_menu_bar()

        imgui.show_test_window()

        imgui.begin("Custom window", True)
        imgui.text("Bar")
        imgui.text_colored("Eggs", 0.2, 1., 0.)
        imgui.end()

        gl.glClearColor(0., 0., 0.2, 1)
        gl.glClear(gl.GL_COLOR_BUFFER_BIT)

        # Here is the glumnpy code inserted in the 
        # pyimgui render loop
        quad['theta'] += 1.0/60.0
        quad.draw(gl.GL_TRIANGLE_STRIP)

        imgui.render()
        impl.render(imgui.get_draw_data())
        glfw.swap_buffers(window)

    impl.shutdown()
    glfw.terminate()


def impl_glfw_init():
    width, height = 1280, 720
    window_name = "minimal ImGui/GLFW3 example"

    if not glfw.init():
        print("Could not initialize OpenGL context")
        exit(1)

    glfw.window_hint(glfw.CONTEXT_VERSION_MAJOR, 3)
    glfw.window_hint(glfw.CONTEXT_VERSION_MINOR, 2)

    # the OPENGL_COMPAT_PROFILE enables the mix between pyimgui and glumpy
    glfw.window_hint(glfw.OPENGL_PROFILE, glfw.OPENGL_COMPAT_PROFILE)
    glfw.window_hint(glfw.OPENGL_FORWARD_COMPAT, gl.GL_TRUE)

    window = glfw.create_window(
        int(width), int(height), window_name, None, None
    )
    glfw.make_context_current(window)

    if not window:
        glfw.terminate()
        print("Could not initialize Window")
        exit(1)

    return window

if __name__ == "__main__":
    main()
```

This is just a proof of concept to show that it is possible to mix glumpy and pyimgui.

pyimgui handles all the events, finding out how to give back some events to glumpy (i.e. window resize) would be the next step.

**Comment by rougier** [comment]
Waouh, fantastic!!! That's very nice.

Could you make a PR with your code ? Maybe with a window allowing to change some actual settings?

**Comment by mlviz** [comment]
I made a branch with a first experimental imgui backend and 2 examples. The only problem is that I needed to modify 2 of the files in the external libraries to make it work, I can find a way to handle that for a build that would be great.

imgui/integrations/opengl.py
we need at line 205 to typecast (I used numpy but  probably would be better to do it with ctype):
```
        ortho_projection = np.array([
            [ 2.0/display_width, 0.0,                   0.0, 0.0],
            [ 0.0,               2.0/-display_height,   0.0, 0.0],
            [ 0.0,               0.0,                  -1.0, 0.0],
            [-1.0,               1.0,                   0.0, 1.0]], dtype=np.float32)
```
and something similar in:

baseplatform.py from PyOpengl.

Otherwise I added the examples in the subdirectory examples/imgui.

![image](https://user-images.githubusercontent.com/45414538/49348808-a5ea4380-f675-11e8-917c-5008d96f2354.png)

I will make a pull request but please validate that you have a version that works, it has been tested on pc.

A+
Martin

**Comment by rougier** [comment]
Thanks. I'll check and test the PR.

**Comment by ivoihrke** [comment]
Hi, is there any update on this ? I think an integration would be a great addition to glumpy. I tested the above code with current pip install versions of imgui and glumpy and it works modulo a simple change in imgui/integrations/opengl.py (line 207), which is the line mentioned above. Imgui uses ctypes already and a change into 

```python
 ortho_projection_list = [
            [ 2.0/display_width, 0.0,                   0.0, 0.0],
            [ 0.0,               2.0/-display_height,   0.0, 0.0],
            [ 0.0,               0.0,                  -1.0, 0.0],
            [-1.0,               1.0,                   0.0, 1.0]
        ];
        
        op_flat_list = [item for sublist in ortho_projection_list for item in sublist]
        ortho_projection = (ctypes.c_float * len(op_flat_list))(*op_flat_list)
```
fixes the issue and makes the example run. I didn't have to touch baseplatform.py in pyopengl.

**Comment by rougier** [comment]
I just forgot about this nice example. It should be definitely integrated into glumpy. Can you make a PR ?

**Comment by ivoihrke** [comment]
I submitted a PR at pyimgui to request the change outlined above. If this is accepted, the demo by mlviz should work out of the box.

https://github.com/swistakm/pyimgui/pull/157

He had submitted a branch as PR for you:

https://github.com/glumpy/glumpy/pull/177

but I cannot find it in the repo. Did you merge this at some point ? It is still marked as open ...
I would be interested in his interaction demos.

**Comment by rougier** [comment]
Right. It's not yet merged. Let's wait for the merge of of your PR then I'll merge #177

**Comment by ivoihrke** [comment]
The fix is merged in pyimgui, a new release 1.2.0 is out that incorporates the change. Thus, 
```
pip install imgui
```
should suffice.
